File: slinding window/max_consecutives_ones3.py
# tc O(n), sc O(1)
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        l = r = res = cur_count = 0
        while r<len(nums):
            if nums[r] == 1:
                r += 1
                cur_count += 1
            elif nums[r] == 0 and k>0:
                k -= 1
                r += 1
                cur_count += 1
            else:
                res = max(res, cur_count)
                while nums[l] == 1:
                    l += 1
                cur_count = r-l
                l += 1
                r += 1
        return max(res,cur_count)

File: slinding window/max_consecutives_ones3.java
public class max_consecutives_ones3 {
    
}
// tc O(n), sc O(1)
class Solution {
    public int longestOnes(int[] nums, int k) {
        int l=0, r=0, cur_count=0, res=0;
        while(r<nums.length){
            if(nums[r]==1){
                r++;
                cur_count++;
            }
            else if(nums[r]==0 && k>0){
                k--;
                cur_count++;
                r++;
            }
            else{
                res=Math.max(res,cur_count);
                while(nums[l]==1){
                    l++;
                }
                l++;
                r++;
                cur_count=r-l;
            }
        }
        return Math.max(res,cur_count);
    }
}